rust_src/src/obarray.rs

```diff
@@ -204,9 +205,10 @@ pub extern "C" fn intern_driver(
 /// A second optional argument specifies the obarray to use;
 /// it defaults to the value of `obarray'.
 #[lisp_fn(min = "1")]
-pub fn intern_soft(name: LispObject, obarray: Option<LispObarrayRef>) -> LispObject {
+pub fn intern_soft(name: LispSymbolOrString, obarray: Option<LispObarrayRef>) -> LispObject {
     let obarray = obarray.unwrap_or_else(LispObarrayRef::global);
     let tem = obarray.lookup(name);
+    let name: LispObject = name.into();
```

Comment:
The only reason to make a LispObject is to answer "is symbol and does my name match". This seems like something the new `LispSymbolOrString` could answer instead of falling back to LispObject.